src/ctest.ts

```diff
@@ -602,14 +604,40 @@ export class CTestDriver implements vscode.Disposable {
         return uniqueTests;
     }
 
-    private async runTestHandler(request: vscode.TestRunRequest, cancellation: vscode.CancellationToken, useCMakePresets: boolean) {
+    // Returns false if test preset wasn't selected already (changing test preset triggers test explorer refresh)
+    private async checkTestPreset(tests: vscode.TestItem[]): Promise<boolean> {
+        let presetMayChange = false;
+        for (const test of tests) {
+            if (test.id === testPresetRequired) {
+                const folder = test.parent ? test.parent.id : test.id;
+                const project = await this.projectController?.getProjectForFolder(folder);
+                if (!project) {
+                    log.error(localize('no.project.found', 'No project found for folder {0}', folder));
+                    return false;
+                }
+                await vscode.commands.executeCommand('cmake.selectTestPreset', project.workspaceFolder);
+                presetMayChange = true;
+            }
+        }
+
+        if (presetMayChange) {
+            return false;
+        }
+        return true;
+    }
+
+    private async runTestHandler(request: vscode.TestRunRequest, useCMakePresets: boolean, cancellation: vscode.CancellationToken) {
         if (!testExplorer) {
             return;
         }
 
         const requestedTests = request.include || this.testItemCollectionToArray(testExplorer.items);
         const tests = this.uniqueTests(requestedTests);
 
+        if (!await this.checkTestPreset(tests)) {
```

Comment:
`useCMakePresets` is passed in to this function, I don't think you should check the test preset if you're not using presets.

Reply:
See the reply to the comment below.